compiler/rustc_codegen_ssa/src/back/link.rs

```diff
@@ -2079,9 +2079,14 @@ fn add_local_native_libraries(
             NativeLibKind::Framework { as_needed } => {
                 cmd.link_framework(name, as_needed.unwrap_or(true))
             }
-            NativeLibKind::Static { whole_archive, .. } => {
+            NativeLibKind::Static { whole_archive, bundle, .. } => {
                 if whole_archive == Some(true)
                     || (whole_archive == None && default_to_whole_archive(sess, crate_type, cmd))
+                    // Backward compatibility case: this can be a rlib (so `+whole-archive` cannot
+                    // be added explicitly if necessary, see the error in `fn link_rlib`) compiled
+                    // as an executable due to `--test`. Use whole-archive implicitly, like before
```

Comment:
Unfortunately we cannot determine whether it would be a rlib or not if `--test` weren't passed, because the rlib-ness is typically specified by cargo.
https://github.com/rust-lang/rust/issues/95561, for example still looks like an executable if we ignore `--test` in `fn collect_crate_types`.